### PR TITLE
Herfindahl-Hirschman Index (HHI)

### PR DESCRIPTION
**Pull request by cfarm**
Adds formula + tests for HHI value. HHI is used in antitrust enforcement to calculate the competitiveness of a market based on the number of firms and their relative market shares.
## Formula

HHI = s1^2 + s2^2 + s3^2 + ... + sn^2 (where sn is the market share of the ith firm).
## Example

"an industry consisting of two firms with market shares of 70% and 30% has an HHI of 70²+30², or 5800."
## Sources

http://www.unclaw.com/chin/teaching/antitrust/herfindahl.htm
https://sites.google.com/site/stuffeconomics/courses/microeconomics/herfindahl-hirschman-index
http://www.investopedia.com/terms/h/hhi.asp
## Additions
- adds formula for calculating HHI value
## Changes
- tests for market share array values
- tests for function call w/o number or args
## Testing
- run `npm test` to see tests pass
## Review
- @ascott1 
- @contolini 

[image](https://www.youtube.com/watch?v=jkRWYC28-Nw)

### CODE REVIEW

**Comment by contolini** [comment]
Thanks, Kowalski.

File: src/herfindahlIndex.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Calculates Herfindahl-Hirschman Index, a measure of market concentration and competition, used by U.S. governments for antitrust enforcement.
+ * @param {array} marketShare (expressed as percentage value for each firm)
+ * @returns {number} Herfindahl-Hirschman Index (HHI), a number between 0 and 10,000
+ */
+
+function herfindahlIndex(marketShare) {
+  var result = 0;
+
+  enforceNumber(marketShare);
```
Comment (contolini) [inline src/herfindahlIndex.js]: Delete this line. `enforceNumber` is used to ensure an argument is a number. In this case we want to ensure it's an array. And the fact that it's not throwing an error indicates it's completely broken...

File: src/herfindahlIndex.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  if (!marketShare) {
```
Comment (contolini) [inline src/herfindahlIndex.js]: And check if it's an array here by changing this condition expression to: `!marketShare || !Array.isArray(marketShare)`.

File: test/herfindahlIndex.js [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
 
-test('calculate futureValue', function(t) {
-  t.plan(1);
+test('calculate Herfindahl Index', function(t) {
+  t.plan(5);
 
+  var calc1 = herfindahlIndex([30, 60]);
 
-// The Herfindahl Index is a measure of the size of firms in relation to the industry and an indicator of the amount of competition among them
+  t.equal(calc1, 4500, 'correctly calculates Herfindahl-Hirschman Index');
```
Comment (contolini) [inline test/herfindahlIndex.js]: Please give each assertion a unique description. e.g. "correctly calculates Herfindahl-Hirschman Index with 2 firms"

File: src/herfindahlIndex.js (hunk 1 above)
Comment (cfarm) [inline src/herfindahlIndex.js]: yeah I wondered. so is the same thing breaking here? https://github.com/CapitalJS/capitaljs/blob/master/src/returnOnInvestment.js#L12

Comment (cfarm) [inline src/herfindahlIndex.js]: looked at this with @ascott1 and it's actually working as expected. `enforceNumber` treats the array arg as an object https://github.com/CapitalJS/capitaljs/blob/master/src/utils/enforce/number/index.js#L7

Comment (contolini) [inline src/herfindahlIndex.js]: Word, it needs to do some type checking.

**Comment by cfarm** [comment]
refactorated without `enforceNumber` - so I could be more specific about argument values in my tests.

**Comment by cfarm** [comment]
I am not to be truffled with.

**Comment by contolini** [comment]
This is all going to H E L L double hockey sticks.

**Pull request merged**